src/api/java/com/minecolonies/api/colony/buildings/IBuilding.java

```diff
@@ -510,10 +510,19 @@ else if (getBuildingLevel() <= WOOD_HUT_LEVEL)
      * Get setting for key. Utility function.
      * @param key the key.
      * @param <T> the key type.
-     * @return the optional wrapping the value.
+     * @return the setting.
      */
     <T extends ISetting> T getSetting(@NotNull final ISettingKey<T> key);
 
+    /**
+     * Get setting for key. Utility function.
+     * @param key the key.
+     * @param <T> the key type.
+     * @return the optional wrapping the value.
+     */
+    @NotNull
+    <T extends ISetting> Optional<T> getOptionalSetting(@NotNull final ISettingKey<T> key);
+
```

Comment:
This was proposed to @Raycoms by me, but he hates this kind of functional API.
Not sure this as such will get accepted ....

Reply:
It's not appropriate in all contexts, but using `Optional`'s `ifPresent`/`map`/`flatMap`/`orElse` interface seems safer to me than surprise nulls, or using a setting that the user can't actually edit (so might not be appropriate)...

There's a bunch of places that have a defensive `building.hasModule(ISettingsModule.class) && building.getSetting(xxx).getValue()`, which is a leaked abstraction (and _still_ potentially vulnerable to nulls).  Would be better (and safer) with `building.getOptionalSetting(xxx).map(StringSetting::getValue).orElse("default")` or similar.

Reply:
I'm just against "getOptionalSetting" in all context. In limited contexts I'm fine with it, it has it's uses, but I don't want to have to rely on it 100%

Reply:
Well, it should probably be used in all contexts where the setting might not exist.  So, any setting with an `isActive` check, or any case where you have a generic building that may or may not have a specific setting.

It's less important in other cases, where you're fairly certain you can't get a null.  But it seems simpler to have a uniform way to query settings even in those cases.

Reply:
i.e. in some cases we do rely on getOptionalModule as well

Reply:
I think we should make this not a getOptionalSetting, but directly a getSettingOrDefault, so we don't require to unpack that on the other side always

Reply:
Well, that could be useful too (particularly in this case), but then you couldn't call other methods on `ISetting`, such as `.ifPresent(setting -> setting.set(foo))`

Reply:
Hrm.  I don't think Java provides any good way to actually write a generic `getSettingOrDefault`.  `ISetting` doesn't provide `getValue`, for a start, nor is there any simple way to extract the value type from the `ISettingKey`.  This probably isn't impossible, but it would require a pretty hefty rewrite.

Reply:
Ah, darn, then :shipit: